src/domain_decomposition.hpp

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <vector>
+
+#include <backends.hpp>
+#include <common_types.hpp>
+#include <communication/global_policy.hpp>
+#include <recipe.hpp>
+#include <util/optional.hpp>
+#include <util/partition.hpp>
+#include <util/transform.hpp>
+
+namespace nest {
+namespace mc {
+
+// Meta data used to guide the domain_decomposition in distributing
+// and grouping cells.
+struct group_rules {
+    cell_gid_type target_group_size;
+    backend_policy policy;
+};
+
+class domain_decomposition {
+public:
+    using gid_type = cell_gid_type;
+
+    struct group_range_type {
+        gid_type from;
+        gid_type to;
+        cell_kind kind;
+    };
+
+    domain_decomposition(const recipe& rec, const group_rules& rules) {
+        EXPECTS(rules.target_group_size>0);
+
+        auto num_domains = communication::global_policy::size();
+        auto domain_id = communication::global_policy::id();
+
+        // partition the cells globally across the domains
+        num_global_cells_ = rec.num_cells();
+        first_cell_ = (gid_type)(num_global_cells_*(domain_id/(double)num_domains));
+        last_cell_ = (gid_type)(num_global_cells_*((domain_id+1)/(double)num_domains));
+
+        // partition the local cells into cell groups
+        if (num_local_cells()>0) {
+            gid_type group_size = 1;
+            group_starts_.push_back(first_cell_);
+            auto group_kind = rec.get_cell_kind(first_cell_);
+            group_kinds_.push_back(group_kind);
+            gid_type gid = first_cell_+1;
```

Comment:
Is +1 really a magic number in this context? It is simply the next index.
I have added a comment that describes the algorithm in more detail, which hopefully makes these steps clearer.